docs/proposals/memory-management.md

```diff
@@ -0,0 +1,135 @@
+
+ There is no container cleanup policy in the event of an OOM kill. What should happen to a pod if a process in one of its container's was OOM killed? 
+
+* Kubelet cannot associate a system memory pressure OOM kill to a specific container or Pod. 
+
+* Memory usage of system daemons is unrestricted. This results in dynamic memory availability on the node. 
+
+* All burstable jobs that exceed their request is equally likely to be killed. Ideally, the job that exceeds the most should be killed. 
+
+* There are no limits placed on kernel resources consumed by jobs (proc, fd, sockets, network buffers, etc)
+
+#####Ideas for improving memory management
+
+1. Define OOM handling policy for containers and Pods:
+
+    *Suggestion*: Whenever an OOM kill occurs in a container, mark that entire container as OOM killed and terminate it. We can provide a knob to alter this behavior if required
+    If a Pod has an app container in "guaranteed" class and a "side-car" container in "Burstable" class, what happens to the pod when the "side-car" is OOM killed? Will the Pod continue to run without the "side-car"?
+    Requiring all containers in a Pod to be of the same QoS class will be simpler.
+
+2. Impose limits on "best effort" and "Burstable" jobs
+
+   *Suggestion*: Don't overcommit memory assigned to "guaranteed" tasks.
+   Use the remaining memory for "Burstable" and best-effort jobs. Whenever there is memory pressure in the "Burstable" + "best effort" bucket, a "best effort" task will most likely be killed based on our QoS policy. 
+
+3. Restrict resource usage of system daemons
+
+   Allocate only the remaining memory to jobs (reserve some memory for the kernel).
+   Even if this might be wasteful, this results in reliable performance. Memory leaks in system daemons are supposed to be caught and fixed before being deployed in production.
+
+4. Set soft limits for Burstable jobs to their memory request: This should result in "Burstable" jobs being pushed to their requested memory limit under system memory pressure. 
+
+5. Kubelet reduces the likelihood of system memory pressure:
+
+  1. Kubelet can subscribe to memory pressure notifications from the kernel, and possibly evict low priority jobs (or)
+  2. Kubelet can set thresholds based notifications on the root cgroup and evict containers that are most suitable candidates based on the QoS policy and current usage. Kubelet can try to limit total memory usage to under 90% of the memory capacity.
+
+6. Eviction policy
+
+   System level job SLOs have to be defined to come up with an eviction policy at the node whenever there is system memory pressure.
+
+7. Disable OOM Kills on containers
+
+   OOM kills can be disabled on all containers. Kubelet (or another daemon) can watch for OOM kill notifications and evict containers based on QoS policy.
+   Disabling OOM kills at the system level will result in a system panic. If kubelet is un-responsive for any reason, we do not want the system to panic. 
```

Comment:
Are these 2 different options? It's not clear whether there's a viable option here.

This part could also use more background, such as ensuring that containers using less than their requests don't receive OOM in the case of system OOM, where oom_score_adj is inadequate because....